File: front/src/ui-runtime/uiSurfaceCatalog.ts
type RouteAliasRule = {
  canonicalPrefix: string;
  aliasPrefix: string;
  exact?: boolean;
};

const ROUTE_ALIAS_RULES: RouteAliasRule[] = [
  { canonicalPrefix: "/characters", aliasPrefix: "/character" },
  { canonicalPrefix: "/workshop", aliasPrefix: "/creative-workshop" },
  { canonicalPrefix: "/channels/wechat", aliasPrefix: "/wechat", exact: true },
  { canonicalPrefix: "/channels/qq", aliasPrefix: "/qq", exact: true },
  { canonicalPrefix: "/settings/devices", aliasPrefix: "/devices" },
  { canonicalPrefix: "/developer/kernel", aliasPrefix: "/kernel" },
  { canonicalPrefix: "/memory/manager", aliasPrefix: "/memory-manager", exact: true },
  { canonicalPrefix: "/memory/timeline", aliasPrefix: "/memory-timeline", exact: true },
  { canonicalPrefix: "/memory/episodic", aliasPrefix: "/episodic", exact: true },
  { canonicalPrefix: "/memory/graph", aliasPrefix: "/graph", exact: true },
  { canonicalPrefix: "/memory/profiles", aliasPrefix: "/profiles", exact: true },
  { canonicalPrefix: "/memory/world-book", aliasPrefix: "/world-book", exact: true },
  { canonicalPrefix: "/chat-logs", aliasPrefix: "/logs", exact: true },
  { canonicalPrefix: "/chat-import", aliasPrefix: "/import", exact: true },
  { canonicalPrefix: "/settings/runtime-mode", aliasPrefix: "/runtime-mode", exact: true },
  { canonicalPrefix: "/settings/long-running", aliasPrefix: "/long-running", exact: true },
  { canonicalPrefix: "/settings/decision-viz", aliasPrefix: "/decision-viz", exact: true },
  { canonicalPrefix: "/settings/privacy-scan", aliasPrefix: "/privacy-scan", exact: true },
  { canonicalPrefix: "/settings/storage", aliasPrefix: "/storage", exact: true },
  { canonicalPrefix: "/settings/user", aliasPrefix: "/user-settings", exact: true },
];

export function normalizeUIRoute(raw: string): string {
  const route = (raw || "/").split(/[?#]/, 1)[0].trim() || "/";
  const prefixed = route.startsWith("/") ? route : `/${route}`;
  return prefixed.length > 1 ? prefixed.replace(/\/+$/, "") : prefixed;
}

function matchesPrefix(route: string, prefix: string, exact = false): boolean {
  return exact ? route === prefix : route === prefix || route.startsWith(`${prefix}/`);
}

function swapPrefix(route: string, from: string, to: string, exact = false): string {
  return exact ? to : `${to}${route.slice(from.length)}`;
}

export function uiRouteAliases(rawRoute: string): string[] {
  const route = normalizeUIRoute(rawRoute);
  const aliases = new Set<string>([route]);
  for (const rule of ROUTE_ALIAS_RULES) {
    if (matchesPrefix(route, rule.canonicalPrefix, rule.exact)) {
      aliases.add(swapPrefix(route, rule.canonicalPrefix, rule.aliasPrefix, rule.exact));
    }
    if (matchesPrefix(route, rule.aliasPrefix, rule.exact)) {
      aliases.add(swapPrefix(route, rule.aliasPrefix, rule.canonicalPrefix, rule.exact));
    }
  }
  return [...aliases];
}

export function canonicalUISurfaceId(rawRoute: string): string {
  const aliases = uiRouteAliases(rawRoute);
  const hasPrefix = (prefix: string) => aliases.some((route) => route === prefix || route.startsWith(`${prefix}/`));
  const hasExact = (...paths: string[]) => paths.some((path) => aliases.includes(path));

  if (hasExact("/characters", "/character")) return "surface.character.list";
  if (hasPrefix("/characters") || hasPrefix("/character")) return "surface.character.detail";
  if (hasExact("/memory/manager", "/memory-manager")) return "surface.memory.manager";
  if (hasPrefix("/memory") || hasExact("/memory-timeline", "/episodic", "/graph", "/profiles", "/world-book")) {
    return "surface.memory.detail";
  }
  if (hasPrefix("/workshop") || hasPrefix("/creative-workshop")) return "surface.workshop";
  if (hasExact("/channels/wechat", "/wechat")) return "surface.channel.wechat";
  if (hasExact("/channels/qq", "/qq")) return "surface.channel.qq";
  if (hasPrefix("/settings/devices") || hasPrefix("/devices")) return "surface.settings.devices";
  if (hasPrefix("/developer/kernel") || hasPrefix("/kernel")) return "surface.kernel";
  if (hasPrefix("/settings")) return "surface.settings.section";
  if (hasPrefix("/extensions") || hasPrefix("/extension/page")) return "surface.extension";
  return "surface.page";
}

File: .baseline-check/mobile_app/lib/core/ui_runtime/routing/ui_surface_catalog.dart
class _RouteAliasRule {
  const _RouteAliasRule(this.canonicalPrefix, this.aliasPrefix, {this.exact = false});

  final String canonicalPrefix;
  final String aliasPrefix;
  final bool exact;

  bool matches(String route, String prefix) {
    if (exact) return route == prefix;
    return route == prefix || route.startsWith('$prefix/');
  }

  String swap(String route, String from, String to) {
    if (exact) return to;
    return '$to${route.substring(from.length)}';
  }
}

const _routeAliasRules = <_RouteAliasRule>[
  _RouteAliasRule('/characters', '/character'),
  _RouteAliasRule('/workshop', '/creative-workshop'),
  _RouteAliasRule('/channels/wechat', '/wechat', exact: true),
  _RouteAliasRule('/channels/qq', '/qq', exact: true),
  _RouteAliasRule('/settings/devices', '/devices'),
  _RouteAliasRule('/developer/kernel', '/kernel'),
  _RouteAliasRule('/memory/manager', '/memory-manager', exact: true),
  _RouteAliasRule('/memory/timeline', '/memory-timeline', exact: true),
  _RouteAliasRule('/memory/episodic', '/episodic', exact: true),
  _RouteAliasRule('/memory/graph', '/graph', exact: true),
  _RouteAliasRule('/memory/profiles', '/profiles', exact: true),
  _RouteAliasRule('/memory/world-book', '/world-book', exact: true),
  _RouteAliasRule('/chat-logs', '/logs', exact: true),
  _RouteAliasRule('/chat-import', '/import', exact: true),
  _RouteAliasRule('/settings/runtime-mode', '/runtime-mode', exact: true),
  _RouteAliasRule('/settings/long-running', '/long-running', exact: true),
  _RouteAliasRule('/settings/decision-viz', '/decision-viz', exact: true),
  _RouteAliasRule('/settings/privacy-scan', '/privacy-scan', exact: true),
  _RouteAliasRule('/settings/storage', '/storage', exact: true),
  _RouteAliasRule('/settings/user', '/user-settings', exact: true),
];

String normalizeUIRoute(String raw) {
  final queryIndex = raw.indexOf('?');
  final hashIndex = raw.indexOf('#');
  var end = raw.length;
  if (queryIndex >= 0 && queryIndex < end) end = queryIndex;
  if (hashIndex >= 0 && hashIndex < end) end = hashIndex;
  var route = raw.substring(0, end).trim();
  if (route.isEmpty) return '/';
  if (!route.startsWith('/')) route = '/$route';
  while (route.length > 1 && route.endsWith('/')) {
    route = route.substring(0, route.length - 1);
  }
  return route;
}

Set<String> uiRouteAliases(String rawRoute) {
  final route = normalizeUIRoute(rawRoute);
  final aliases = <String>{route};
  for (final rule in _routeAliasRules) {
    if (rule.matches(route, rule.canonicalPrefix)) {
      aliases.add(rule.swap(route, rule.canonicalPrefix, rule.aliasPrefix));
    }
    if (rule.matches(route, rule.aliasPrefix)) {
      aliases.add(rule.swap(route, rule.aliasPrefix, rule.canonicalPrefix));
    }
  }
  return aliases;
}

String canonicalUISurfaceId(String rawRoute) {
  final aliases = uiRouteAliases(rawRoute);
  bool hasPrefix(String prefix) => aliases.any(
        (route) => route == prefix || route.startsWith('$prefix/'),
      );

  if (aliases.contains('/characters') || aliases.contains('/character')) {
    return 'surface.character.list';
  }
  if (hasPrefix('/characters') || hasPrefix('/character')) {
    return 'surface.character.detail';
  }
  if (aliases.contains('/memory/manager') || aliases.contains('/memory-manager')) {
    return 'surface.memory.manager';
  }
  if (hasPrefix('/memory') ||
      aliases.any((route) => const {
            '/memory-timeline',
            '/episodic',
            '/graph',
            '/profiles',
            '/world-book',
          }.contains(route))) {
    return 'surface.memory.detail';
  }
  if (hasPrefix('/workshop') || hasPrefix('/creative-workshop')) {
    return 'surface.workshop';
  }
  if (aliases.contains('/channels/wechat') || aliases.contains('/wechat')) {
    return 'surface.channel.wechat';
  }
  if (aliases.contains('/channels/qq') || aliases.contains('/qq')) {
    return 'surface.channel.qq';
  }
  if (hasPrefix('/settings/devices') || hasPrefix('/devices')) {
    return 'surface.settings.devices';
  }
  if (hasPrefix('/developer/kernel') || hasPrefix('/kernel')) {
    return 'surface.kernel';
  }
  if (hasPrefix('/settings')) return 'surface.settings.section';
  if (hasPrefix('/extensions') || hasPrefix('/extension/page')) {
    return 'surface.extension';
  }
  return 'surface.page';
}
